python/tskit/combinatorics.py

```diff
@@ -28,12 +28,133 @@
+        for label in leaf_labels[1:]:
+            n = TreeNode(label=label)
+            x = rng.choice(nodes)
+            internal = TreeNode(parent=x.parent, children=[x, n])
+            if x.parent is not None:
+                index = x.parent.children.index(x)
+                x.parent.children[index] = internal
+            rng.shuffle(internal.children)
+            x.parent = internal
+            n.parent = internal
+            nodes.extend([n, internal])
+
+        root = nodes[0]
+        while root.parent is not None:
+            root = root.parent
+        return root
+
+
+def split_polytomies(
+    tree,
+    *,
+    epsilon=None,
+    method=None,
+    record_provenance=True,
+    random_seed=None,
+):
+    """
+    Return a new tree where extra nodes and edges have been inserted
+    so that any any node with more than two children is resolved into
+    a binary tree.
+
+    For further documentation, please refer to the :meth:`Tree.split_polytomies`
+    method, which is the usual route through which this function is called.
+    """
+    allowed_methods = ["random"]
+    if epsilon is None:
+        epsilon = 1e-10
+    if method is None:
+        method = "random"
+    if method not in allowed_methods:
+        raise ValueError(f"Method must be chosen from {allowed_methods}")
+
+    tables = tree.tree_sequence.dump_tables()
+    tables.keep_intervals([tree.interval], simplify=False)
+    tables.edges.clear()
+    rng = random.Random(random_seed)
+
+    for u in tree.nodes():
+        if tree.num_children(u) > 2:
+            root = TreeNode.random_binary_tree(tree.children(u), rng)
+            # TODO we should do better here with epsilon. We can easily
+            # split the paths into equally spaced chunks based on the
+            # time of the children using a traversal. We could also do
+            # this is there's mutations, we just take the time of the
+            # node to be the time of the oldest branch above it.
+            root.label = u
+            time = tree.time(u) - epsilon
+            stack = [(child, time) for child in root.children]
+            while len(stack) > 0:
+                node, time = stack.pop()
+                if node.label is None:
+                    node.label = tables.nodes.add_row(time=time)
+                tables.edges.add_row(*tree.interval, node.parent.label, node.label)
+                for child in node.children:
+                    stack.append((child, time - epsilon))
+        else:
+            for v in tree.children(u):
+                tables.edges.add_row(*tree.interval, u, v)
+
+    if record_provenance:
+        parameters = {"command": "split_polytomies"}
+        tables.provenances.add_row(
+            record=json.dumps(tskit.provenance.get_provenance_dict(parameters))
+        )
+    try:
+        tables.sort()
+        ts = tables.tree_sequence()
+    except tskit.LibraryError as e:
+        msg = str(e)
+        if msg.startswith(
```

Comment:
If we are checking on the time of children or mutations above children, then I think it would be better to raise these errors when we know we are going to violate the time constraints, as then we can say which nodes or mutations are causing the problems, which is much more helpful to the user.

I didn't do this for mutation times because I removed the code for checking mutation times, but if we are going to reinstate it, we should raise the error there.